### ISSUE TITLE
ETIMEDOUT unless `?ssl=1` explicitly set - difference in behaviour since v4?

### ISSUE THREAD

**Issue by stefan1968** [graphile/postgraphile#708]
I'm submitting a ... 

* [x ] bug report
* [ ] feature request
* [ ] question 

PostGraphile version:4.0.0.-beta.2

Minimal SQL file that can be loaded into a clean database:

```sql

```

Steps to reproduce:

```javascript
const express = require('express');
const { postgraphile } = require('postgraphile');
const config = require('./config/index');

const app = express();


app.use(postgraphile(config.postgraphile.host, config.postgraphile.schema_name, {
    pgDefaultRole: config.postgraphile.default_role,
    enableCors: true,
    jwtSecret: config.postgraphile.jwt_secret,
    jwtPgTypeIdentifier: config.postgraphile.jwt_token,
    graphiql: config.postgraphile.graphiql,
    jwtVerifyOptions: {
        ignoreExpiration: true

    }
}));

app.listen(config.postgraphile.port || 5000);
```

This times out and refuses to connect, can't fetch schema etc.

Error: connect ETIMEDOUT 34.234.132.151:5432
    at Object._errnoException (util.js:1024:11)
    at _exceptionWithHostPort (util.js:1046:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1182:14)

Change it to postgraphql, works fine. 
i.e const { postgraphql } = require('postgraphql') works fine.

When i did sort of get it working, the authenticate example kept throwing up the error 'ROW EXPANSION VIA * NOT ALLOWED HERE'

So, currently can't use the authenticate example on postgraphile,  and can't use postgraphile in express.

Current behavior: Times out

Expected behavior: Working express version (works fine if use older postgraphql)

**Comment by benjie** [comment]
Do you have environmental variables set for your database? PostGraphile sources additional information from the environment versus v3. The error seems to be that the database you’re connecting to is on host 34.234.132.151 and the connection is timing out. Can you confirm that you can connect to that host and that port number via “psql”? I’m guessing that PostGraphQL is just connecting to your localhost database hence the lack of issue with that.  Can you confirm the shape of config.postgraphile.host?

**Comment by stefan1968** [comment]
Environment variables are set, the db is remote (Amazon RDS Instance). Postgraphql connects just fine. psql connects just fine.  Like I said change to postgraphql it works. That's the *only* change, no environment change, no config change.

```javascript
const express = require('express');
const { postgraphql } = require('postgraphql');
const config = require('./config/index');

const app = express();


app.use(postgraphql(config.postgraphile.host, config.postgraphile.schema_name, {
    pgDefaultRole: config.postgraphile.default_role,
    enableCors: true,
    jwtSecret: config.postgraphile.jwt_secret,
    jwtPgTypeIdentifier: config.postgraphile.jwt_token,
    graphiql: config.postgraphile.graphiql,
    jwtVerifyOptions: {
        ignoreExpiration: true

    }
}));

app.listen(config.postgraphile.port || 5000);
```

**Comment by benjie** [comment]
I can assure you that I successfully use PostGraphile beta successfully with many different express apps, most of which use RDS, as do many other users. There must be something specific to your setup that is triggering the issue; we need to get to the bottom of what that is.

You say "Environment variables are set" - can you confirm which envvars they are? Are there any beginning `PG` or `DATABASE_`?

The host, being RDS, most likely has a firewall which only allows certain IP addresses to connect to it; but given when you change the library to postgraphql it suddenly works we can rule that out (assuming that they both run on the same host).

That said, the error itself is an `ETIMEDOUT` error - which implies that the connection to RDS is not going through. Can you try appending `?ssl=1` to the connection string to enforce SSL? Perhaps this is the issue - although I wouldn't expect that to result in an `ETIMEDOUT`.

Can you confirm you're not running any of this in a VM or container or remote server that may be messing with DNS resolution?

Have you tried connecting to a local PostgreSQL database rather than an RDS instance?

**Comment by stefan1968** [comment]
"server": {
        "port": 3001
    },
    "postgraphile": {
        "host": "postgres://origin_postgraphql:xxxxxxx@origingg.cjdraitfnk0j.us-east-1.rds.amazonaws.com:5432/originGG",
        "default_role": "origin_anonymous",
        "schema_name": "origin",
        "jwt_secret": "bbo9Q4jsIkfQ1gkpolQAKLXO4WZ-s3SEcvo3gHwfxCEM_IBSisDWzlwcmDKjVfH0",
        "jwt_token": "origin.jwt_token",
        "graphiql": false,
        "port": 5000
    },
    "graphql" : {
        "graphql_url": "http:localhost:5000/graphql",
        "graphiql_url": "http:localhost:5000/graphiql"
    }
}

That is the config json file that works with postgraphql. I'll try adding ssl=1

**Comment by benjie** [comment]
You've just shared your JWT secret; you're going to want to change that now.

**Comment by stefan1968** [comment]
Ha. I will (It's just random string), but yes the ssl=1 fixed the issue. That should be highlighted more in the documentation.
We can close this now, unless anyone wants to add anything.

**Comment by benjie** [comment]
Wow; I was honestly not expecting that to fix it. I'm going to have to dig through the code and figure out why that default has changed now! Please can you confirm the versions of the `pg` and `pg-connection-string` modules that `postgraphql` and `postgraphile` are using please - have a poke in your node_modules folder. Or if you have a lockfile you could just send that 👍

**Comment by benjie** [comment]
I'm repurposing the issue 👍

**Comment by stefan1968** [comment]
Sent lockfile. Really appreciate your swift response.

**Comment by stefan1968** [comment]
Ok, now thats working, we're back to the other problem with the authenticate function.
This function works if i use postgraphql.

With extended errors config, i get the following:-

```
{
  "errors": [
    {
      "errcode": "0A000",
      "message": "row expansion via \"*\" is not supported here",
      "locations": [
        {
          "line": 29,
          "column": 3
        }
      ],
      "path": [
        "authenticate"
      ],
      "stack": "error: row expansion via \"*\" is not supported here\n    at Connection.parseE (/Users/stefanwalker/Projects/Origin.gg/new_backend/graphql/node_modules/pg/lib/connection.js:545:11)\n    at Connection.parseMessage (/Users/stefanwalker/Projects/Origin.gg/new_backend/graphql/node_modules/pg/lib/connection.js:370:19)\n    at TLSSocket.<anonymous> (/Users/stefanwalker/Projects/Origin.gg/new_backend/graphql/node_modules/pg/lib/connection.js:113:22)\n    at emitOne (events.js:116:13)\n    at TLSSocket.emit (events.js:211:7)\n    at addChunk (_stream_readable.js:263:12)\n    at readableAddChunk (_stream_readable.js:250:11)\n    at TLSSocket.Readable.push (_stream_readable.js:208:10)\n    at TLSWrap.onread (net.js:594:20)"
    }
  ],
  "data": {
    "authenticate": null
  }
}
```

**Comment by benjie** [comment]
What do you get if you run `echo $PGSSLMODE` ? It seems ssl has been default off in `pg` since forever (5 years ago). I think pg@6 is negotiating SSL in a different way to pg@7.

**Comment by stefan1968** [comment]
It's not defined. (At least locally)

**Comment by benjie** [comment]
@stefan1968 Can you run your system with the `DEBUG="*:sql"` envvar set (e.g. `export DEBUG="*:sql"; npm run start`) and then send the SQL query that's being generated that is causing this error?

**Comment by stefan1968** [comment]
Sure:-
```
graphile-build-pg:sql
  graphile-build-pg:sql       with __local_0__ as (
  graphile-build-pg:sql         select __local_1__.* from "origin"."authenticate"($1, $2) __local_1__
  graphile-build-pg:sql       )
  graphile-build-pg:sql       select (__local_0__)::text from __local_0__ +0ms
  graphile-build-pg:sql
  graphile-build-pg:sql       with __local_0__ as (
  graphile-build-pg:sql         select (str::"origin"."auth_payload").*
  graphile-build-pg:sql         from unnest(($1)::text[]) str
  graphile-build-pg:sql       )
  graphile-build-pg:sql
  graphile-build-pg:sql       select to_json((case when (__local_0__."jwt_token") is not distinct from null then null else to_json((__local_0__."jwt_token").*) end)) as "jwtToken", to_json((__local_0__."organisation")) as "organisation"
  graphile-build-pg:sql       from __local_0__ as __local_0__
  graphile-build-pg:sql
  graphile-build-pg:sql       where (not (__local_0__ is null)) and (TRUE) and (TRUE)
  graphile-build-pg:sql
  graphile-build-pg:sql
  graphile-build-pg:sql
  graphile-build-pg:sql      +238ms
1 error(s) as origin_anonymous in 1201.07ms :: mutation authenticate { authenticate(input: {email: "stefan.walker@gmail.com", password: "buzz1968"}) { resultData: authPayload { jwtToken organisation } } } query getBlogs($subDomain: String!) { allSponsors(condition: {organisation: $subDomain}) { edges { node { linkUrl imageUrl createdAt updatedAt } } } } query getTheme($subDomain: String!) { resultData: themeByThemeName(themeName: $subDomain) { themeData themeStructure themeName } } query getOrg($subDomain: String!) { resultData: organisationAccountBySubDomain(subDomain: $subDomain) { name description fbLink twitterLink twitchLink logo subDomain primaryColor twitterFeedUsername themesByThemeName { edges { node { themeName themeData themeStructure } } } } } query recentMatches($organisation: String!) { allRecentmatches(condition: {organisation: $organisation}) { totalCount edges { node { id organisation oppositeTeamName oppositeTeamLogo gameName gameLogo score } } } }
```

**Comment by benjie** [comment]
What version of PostgreSQL are you using?

**Comment by stefan1968** [comment]
PostgreSQL 9.6.5

**Comment by benjie** [comment]
Reproduced 👍

**Comment by benjie** [comment]
Fixed in https://github.com/graphile/graphile-build/pull/196

**Comment by benjie** [comment]
Fixed in postgraphile@4.0.0-beta.3
